{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import sys\n",
    "from glob import glob\n",
    "import re\n",
    "from pysam import VariantFile, VariantRecord\n",
    "from collections import defaultdict\n",
    "import csv \n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Download the latest files from the mutation catalogue github repo\n",
    "\n",
    "The excel file contains the graded mutations and their confidence levels\n",
    "The vcf file contains the genomic coordinates of the mutations linked to the graded mutations\n",
    "\"\"\"\n",
    "\n",
    "excel_file_url = 'https://github.com/GTB-tbsequencing/mutation-catalogue-2023/raw/main/Final%20Result%20Files/WHO-UCN-TB-2023.6-eng.xlsx'\n",
    "r = requests.get(excel_file_url, allow_redirects=True)\n",
    "excel_file = 'WHO-UCN-TB-2023.5-eng.xlsx'\n",
    "open(excel_file, 'wb').write(r.content)\n",
    "\n",
    "vcf_file_url = 'https://github.com/GTB-tbsequencing/mutation-catalogue-2023/raw/main/Final%20Result%20Files/Genomic_coordinates_2Feb2024.vcf.gz'\n",
    "r = requests.get(vcf_file_url, allow_redirects=True)\n",
    "open('Genomic_coordinates_7Sep2023.vcf.gz', 'wb').write(r.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "snpEff doesn't use the right codon table for Mtb, so we need to modify the config file\n",
    "\"\"\"\n",
    "snpeff_dirs = glob(f'{sys.prefix}/share/snpeff*')\n",
    "if len(snpeff_dirs)==1:\n",
    "    snpeff_dir = snpeff_dirs[0]\n",
    "    snpeff_config = f'{snpeff_dir}/snpEff.config'\n",
    "    if \"Mycobacterium_tuberculosis_h37rv.Chromosome.codonTable : Bacterial_and_Plant_Plastid\" not in open(snpeff_config).read():\n",
    "        open(snpeff_config, 'a').write(\"\\nMycobacterium_tuberculosis_h37rv.Chromosome.codonTable : Bacterial_and_Plant_Plastid\\n\")\n",
    "    else:\n",
    "        print(\"Right codon table alread set\")\n",
    "else:\n",
    "    print(\"Multiple or no snpeff installations found. Please remove all but one.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "We annotate the vcf file with snpEff to make sure we are getting the same annotations as the graded mutations\n",
    "\"\"\"\n",
    "if os.path.isfile('ann.vcf'):\n",
    "    print(\"Annotation file already exists\")\n",
    "else:\n",
    "    sp.call(\"bcftools view Genomic_coordinates_7Sep2023.vcf.gz | sed 's/NC_000962.3/Chromosome/' | snpEff ann Mycobacterium_tuberculosis_h37rv  > ann.vcf\", shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "This function extracts the both the graded mutation and the snpEff mutation from the vcf file\n",
    "\"\"\"\n",
    "def extract_mutation(var: VariantRecord) -> dict:\n",
    "    nucleotide_types = set(['initiator_codon_variant','synonymous_variant','upstream_gene_variant','splice_region_variant&stop_retained_variant','splice_region_variant&non_coding_transcript_exon_variant','non_coding_transcript_exon_variant'])\n",
    "    gene = var.info['graded_variant'].split(\"_\")[0]\n",
    "    cat_mut = '_'.join(var.info['graded_variant'].split(\"_\")[1:])\n",
    "    annotations = []\n",
    "    for a in var.info['ANN']:\n",
    "        a = a.split('|')\n",
    "        if a[1] in ('downstream_gene_variant'):\n",
    "            continue\n",
    "        if a[3]==gene:\n",
    "            annotations.append(a)\n",
    "    if len(annotations) == 0:\n",
    "        raise Exception\n",
    "    else:\n",
    "        a = annotations[0]\n",
    "        if a[1] in nucleotide_types:\n",
    "            mutation = a[9]\n",
    "        elif a[1]=='start_lost':\n",
    "            mutation = 'p.Met1?'\n",
    "        elif r:=re.match('p.Ter(\\d+)fs',a[10]):\n",
    "            mutation = f'p.Ter{r.group(1)}ext*?'\n",
    "        else:\n",
    "            mutation = a[10]\n",
    "        return {\n",
    "            'gene':gene,\n",
    "            'catalogue_mutation':cat_mut,\n",
    "            'snpEff_mutation':mutation,\n",
    "            'type': a[1]\n",
    "        }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Process the vcf file and compare the graded mutations with the snpEff mutations\n",
    "Some of the mutations don't agree in terms of the assigned graded mutation and the mutation from snpEff so we dump those\n",
    "\"\"\"\n",
    "processed = []\n",
    "dump = []\n",
    "vcf = VariantFile('ann.vcf')\n",
    "for i,var in enumerate(vcf):\n",
    "    try:\n",
    "        obj = extract_mutation(var)\n",
    "    except:\n",
    "        break\n",
    "    if obj['catalogue_mutation']!=obj['snpEff_mutation']:\n",
    "        dump.append((var,obj))\n",
    "    else:\n",
    "        processed.append(obj)\n",
    "\n",
    "\n",
    "print('Number of validated mutations:', len(processed))\n",
    "print('Number of mutations that could not be parsed:', len(dump))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Create a lookup table for the mutations graded mutation -> snpEff mutation\n",
    "This is a one to many relationship. For example the graded mutation katG LoF is linked to many snpEff mutations (e.g. frameshifts)\n",
    "\"\"\"\n",
    "variant_lookup = defaultdict(set)\n",
    "for var in processed:\n",
    "    key = (var['gene'],var['catalogue_mutation'])\n",
    "    val = (var['gene'],var['snpEff_mutation'])\n",
    "    variant_lookup[key].add(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Create function to generate the rows with different SO terms for LoF mutations\n",
    "\"\"\"\n",
    "def generate_lof_rows(gene: str, drug: str, confidence: str, comment: str, vtype: str) -> list:\n",
    "    if vtype=='deletion':\n",
    "        so_terms = ['feature_ablation']\n",
    "    elif vtype=='LoF':\n",
    "        so_terms = ['transcript_ablation','feature_ablation','start_lost','stop_gained','frameshift_variant']\n",
    "    lof_rows = []\n",
    "    for so_term in so_terms:\n",
    "        lof_rows.append({\n",
    "            'Gene':gene,\n",
    "            'Mutation':so_term,\n",
    "            'type': 'drug_resistance' if confidence in ('Assoc w R', 'Assoc w R - Interim') else 'who_confidence',\n",
    "            'drug': drug,\n",
    "            'original_mutation': vtype,\n",
    "            'confidence': confidence,\n",
    "            'source': 'WHO catalogue v2',\n",
    "            'comment': comment\n",
    "        })\n",
    "    return lof_rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\"\n",
    "Here we run through the excel file and lookup the graded mutations in the lookup table.\n",
    "If the mutation is not found in the lookup table we dump it to a json file for checking.\n",
    "If all goes well they should all be found in the lookup table.\n",
    "\"\"\"\n",
    "\n",
    "unparsed = []\n",
    "lof_variants = ['deletion','LoF']\n",
    "confidence = {}\n",
    "rows = []\n",
    "gene_drug_associations = set()\n",
    "for row in pd.read_excel(excel_file,sheet_name='Catalogue_master_file', skiprows=2).to_dict('records'):\n",
    "    if row['gene']=='dnaA' and row['mutation'].startswith('c.-'): \n",
    "        continue\n",
    "    drug = row['drug'].lower()\n",
    "    key = (drug,row['gene'],row['mutation'])\n",
    "    vkey = (row['gene'],row['mutation'])\n",
    "    if row['mutation'] not in lof_variants and vkey not in variant_lookup:\n",
    "        unparsed.append(key)\n",
    "        continue\n",
    "    gene_drug_associations.add((row['gene'], drug))\n",
    "    confidence = row['FINAL CONFIDENCE GRADING'][3:]\n",
    "    for v in variant_lookup[vkey]:\n",
    "        r = {\n",
    "            'Gene':v[0],\n",
    "            'Mutation':v[1],\n",
    "            'type': 'drug_resistance' if confidence in ('Assoc w R', 'Assoc w R - Interim') else 'who_confidence',\n",
    "            'drug': drug,\n",
    "            'original_mutation': row[\"mutation\"],\n",
    "            'confidence': confidence,\n",
    "            'source': 'WHO catalogue v2',\n",
    "            'comment': row[\"Comment\"]\n",
    "        }\n",
    "        rows.append(r)\n",
    "    if row['mutation'] in lof_variants:\n",
    "        if row['mutation']=='deletion':\n",
    "            so_terms = ['feature_ablation']\n",
    "        elif row['mutation']=='LoF':\n",
    "            so_terms = ['transcript_ablation','feature_ablation','start_lost','stop_gained','frameshift_variant']\n",
    "        for so_term in so_terms:\n",
    "            r = {\n",
    "                'Gene':row['gene'],\n",
    "                'Mutation': so_term,\n",
    "                'type': 'drug_resistance' if confidence in ('Assoc w R', 'Assoc w R - Interim') else 'who_confidence',\n",
    "                'drug': drug,\n",
    "                'original_mutation': row[\"mutation\"],\n",
    "                'confidence': confidence,\n",
    "                'source': 'WHO catalogue v2',\n",
    "                'comment': row[\"Comment\"]\n",
    "            }\n",
    "            rows.append(r)\n",
    "\n",
    "print(\"Number of unparsed mutations:\",len(unparsed))\n",
    "print(\"Number of parsed mutations:\",len(rows))\n",
    "\n",
    "import json\n",
    "json.dump(unparsed,open('who.json','w'),indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"This adds pretomanid to the gene-drug associations based on the expert rules\"\"\"\n",
    "for gene in \"ddn, fbiA, fbiB, fbiC, fgd1, Rv2983\".split(\", \"):\n",
    "    rows += generate_lof_rows(gene, 'pretomanid', 'Assoc w R - Interim', 'Confers DLM-PMD cross-resistance', 'LoF')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This adds the expert rule for Rifampicin/RRDR:\n",
    " - Non-silent variants in RRDR of rpoB\n",
    "\"\"\"\n",
    "for mutation in ('missense_variant_p.426_452','conservative_inframe_deletion_c.1276_1356','conservative_inframe_insertion_c.1276_1356'):\n",
    "    rows.append({\n",
    "        'Gene':'rpoB',\n",
    "        'Mutation':mutation,\n",
    "        'type':'drug_resistance',\n",
    "        'drug':'rifampicin',\n",
    "        'original_mutation':'RRDR non-silent',\n",
    "        'confidence':'Assoc w R - Interim',\n",
    "        'source': 'WHO catalogue v2',\n",
    "        'comment':'Expert rule: Non-silent variants in RRDR of rpoB'\n",
    "    }) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Create the mutations output file \n",
    "\"\"\"\n",
    "\n",
    "df = pd.DataFrame(rows)\n",
    "df.to_csv('parsed_who_mutations.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Create a watchlist with all genes, regardless if they have a tier 1/2 mutation or not\n",
    "\"\"\"\n",
    "\n",
    "watchlist_rows = []\n",
    "for gene,drug in gene_drug_associations:\n",
    "    watchlist_rows.append({\n",
    "        'Gene':gene,\n",
    "        'Type':'drug_resistance',\n",
    "        'Drug':drug,\n",
    "    })\n",
    "with open(\"genes.csv\",\"w\")  as O:\n",
    "    writer = csv.DictWriter(O,fieldnames=['Gene','Type','Drug'])\n",
    "    writer.writeheader()\n",
    "    writer.writerows(watchlist_rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "supported_so_terms = [\n",
    "    'coding_sequence_variant', 'chromosome', 'duplication', 'inversion', 'coding_sequence_variant', \n",
    "    'inframe_insertion', 'disruptive_inframe_insertion', 'inframe_deletion', 'disruptive_inframe_deletion', \n",
    "    'downstream_gene_variant', 'exon_variant', 'exon_loss_variant', 'exon_loss_variant', 'duplication', \n",
    "    'duplication', 'inversion', 'inversion', 'frameshift_variant', 'gene_variant', 'feature_ablation', \n",
    "    'duplication', 'gene_fusion', 'gene_fusion', 'bidirectional_gene_fusion', 'rearranged_at_DNA_level', \n",
    "    'intergenic_region', 'conserved_intergenic_variant', 'intragenic_variant', 'intron_variant', \n",
    "    'conserved_intron_variant', 'miRNA', 'missense_variant', 'initiator_codon_variant', 'stop_retained_variant', \n",
    "    'protein_protein_contact', 'structural_interaction_variant', 'rare_amino_acid_variant', \n",
    "    'splice_acceptor_variant', 'splice_donor_variant', 'splice_region_variant', 'splice_region_variant', \n",
    "    'splice_region_variant', 'stop_lost', '5_prime_UTR_premature_', 'start_codon_gain_variant', \n",
    "    'start_lost', 'stop_gained', 'synonymous_variant', 'start_retained', 'stop_retained_variant', \n",
    "    'transcript_variant', 'transcript_ablation', 'regulatory_region_variant', 'upstream_gene_variant', \n",
    "    '3_prime_UTR_variant', '3_prime_UTR_truncation + exon_loss', '5_prime_UTR_variant', \n",
    "    '5_prime_UTR_truncation + exon_loss_variant', 'sequence_feature + exon_loss_variant', 'functionally_normal',\n",
    "    'conservative_inframe_deletion', 'conservative_inframe_insertion'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def so_term_in_mutation(mutation: str) -> bool:\n",
    "    for term in supported_so_terms:\n",
    "        if term in mutation:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "so_term_in_mutation('conservative_inframe_deletion_c.1276_1356')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
